library/core/src/mem/transmutability.rs

```diff
@@ -1,10 +1,80 @@
 use crate::marker::{ConstParamTy_, UnsizedConstParamTy};
 
-/// Are values of a type transmutable into values of another type?
+/// Marks that `Src` is transmutable into `Self`.
 ///
-/// This trait is implemented on-the-fly by the compiler for types `Src` and `Self` when the bits of
-/// any value of type `Self` are safely transmutable into a value of type `Dst`, in a given `Context`,
-/// notwithstanding whatever safety checks you have asked the compiler to [`Assume`] are satisfied.
+/// # Implementation
+///
+/// This trait cannot be implemented explicitly. It is implemented on-the-fly by
+/// the compiler for all types `Src` and `Self` such that, given a set of safety
+/// obligations on the programmer (see [`Assume`]), the compiler has proved that
+/// the bits of a value of type `Src` can be soundly reinterpreted as a `Self`.
+///
+/// Specifically, this trait models (and
+/// [implements](BikeshedIntrinsicFrom::transmute)) the semantics of
+/// transmute-via-union; i.e.:
+///
+/// ```rust
+/// pub unsafe fn transmute_via_union<Src, Dst>(src: Src) -> Dst {
+///     use core::mem::ManuallyDrop;
+///
+///     #[repr(C)]
+///     union Transmute<Src, Dst> {
+///         src: ManuallyDrop<Src>,
+///         dst: ManuallyDrop<Dst>,
+///     }
+///
+///     let transmute = Transmute {
+///         src: ManuallyDrop::new(src),
+///     };
+///
+///     let dst = transmute.dst;
+///
+///     ManuallyDrop::into_inner(dst)
+/// }
+/// ```
+///
+/// Note that this construction supports some transmutations that are unsound
+/// with [`mem::transmute_copy`](super::transmute_copy); namely, transmutations
+/// that extend the bits of `Src` with trailing padding to fill trailing
+/// uninitialized bytes of `Self`; e.g.:
+///
+/// ```ignore (cannot doctest this until bootstrap; remove ignore and corresponding test in tests/mem.rs in the future)
```

Comment:
I believe you can use doc comments' desugared form to gate this correctly.

```suggestion
#[cfg_attr(bootstrap), doc = "```rust,ignore")]
#[cfg_attr(not(bootstrap)), doc = "```rust")]
```

Reply:
Neat! [Done!](https://github.com/rust-lang/rust/compare/7254d9d761afb649b646f16a469683c4ebc47f9a..60cfa88d3f6f37ae0ce2ebd4ac4ece3aa9538a27)